src/runtime/container.cc

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * \file src/runtime/container.cc
+ * \brief POD container type implementations.
+ */
+#include <dmlc/logging.h>
+#include <tvm/runtime/container.h>
+#include <cstdint>
+#include "object_internal.h"
+#include "runtime_base.h"
+
+namespace tvm {
+namespace runtime {
+
+ObjectRef* ADTObj::fields() const {
+  if (this->size_ <= 0) {
+    return nullptr;
+  }
+  auto p = reinterpret_cast<uint32_t*>(const_cast<ADTObj*>(this)) + 2;
+  return reinterpret_cast<ObjectRef*>(p);
+}
+
+ADT::ADT(uint32_t tag, std::vector<ObjectRef> fields) {
+  auto ptr = make_object_with_extra_space<ADTObj>(fields.size() * sizeof(ObjectRef));
+  ptr->tag_ = tag;
+  ptr->size_ = fields.size();
+  for (size_t i = 0; i < fields.size(); ++i) {
+    ptr->fields()[i] = fields[i];
```

Comment:
Direct assignment in here might case a problem. What if the original location ObjectRef is not initialized to nullptr? You will need to first fill the memory with nullptr before running copy. Or use the inplace memory new

Reply:
Will do. Seems like we should also delete the assignment operator of `ObjectRef` so we can get compile time error early.